---
output:
  html_document:
    theme: spacelab
---

<a href="https://jhubiostatistics.shinyapps.io/recount/"><img src="https://raw.githubusercontent.com/leekgroup/recount-website/master/website/recount.png" align="center"></a>

# recount workshop: Learn to leverage 70,000 human RNA-seq samples for your projects

This workshop is for the [BioC2017 meeting](http://bioconductor.org/help/course-materials/2017/BioC2017/).

## Level of difficulty

beginner

## Abstract

The recount2 project re-processed RNA sequencing (RNA-seq) data on over 70,000 human RNA-seq samples spanning a variety of tissues, cell types and disease conditions. Researchers can easily access these data via the `recount` Bioconductor package, and can quickly import gene, exon, exon-exon junction and base-pair coverage data for uniformly processed data from the SRA, GTEx and TCGA projects in R for analysis. This workshop will cover different use cases of the `recount` package, including downloading and normalizing data, processing and cleaning relevant phenotype data, performing differential expression (DE) analyses, and creating reports for exploring the results using other Bioconductor packages. The workshop will also cover how to use the base-pair coverage data for annotation-agnostic DE analyses and for visualizing coverage data for features of interest. After taking this workshop, attendees will be ready to enhance their analyses by leveraging RNA-seq data from 70,000 human samples.

[recount2 website](https://jhubiostatistics.shinyapps.io/recount/), [recount package](http://bioconductor.org/packages/recount), [paper](http://www.nature.com/nbt/journal/v35/n4/full/nbt.3838.html).

## Topic

RNA-seq

## Expected outcomes

Learn how to search projects, download data, explore the metadata, add more phenotype information, and prepare the data for a DE analysis. Then perform the DE analysis with `DESeq2` and explore the results using `regionReport`.
Participant prerequisites: basic familiarity with packages such as `GenomicRanges` and `DESeq2`. Functions from those packages used in the workshop will be briefly described.

## Installation

```{r, eval = FALSE}
## You will need R 3.4.x and Bioconductor 3.6 (currently Bioc-devel)
library('devtools')
install_github('LieberInstitute/recountWorkshop')
```

## Main workshop file

* [recount workshop](compiled_vignette/recount-workshop.html) main file, [R](compiled_vignette/recount-workshop.R) code, raw [Rmd](compiled_vignette/recount-workshop.Rmd) file

### Report files created in the workshop

* [SRP056604_main-results](compiled_vignette/SRP056604_main-results.html) report, [bibliography](compiled_vignette/SRP056604_main-results.bib)
* [SRP056604_sex-predictions](compiled_vignette/SRP056604_sex-predictions.html) report, [bibliography](compiled_vignette/SRP056604_sex-predictions.bib)
* [SRP056604_DERs](compiled_vignette/SRP056604_DERs.html) report, [bibliography](compiled_vignette/SRP056604_DERs.bib)

## Feedback and bug reports

For feedback and bug reports, please create a [new issue](https://github.com/LieberInstitute/recountWorkshop/issues). Remember to check the [posting guidelines](http://www.bioconductor.org/help/support/posting-guide/) from the Bioconductor support website before submitting your new post. Thank you.



Last updated: `r Sys.time()`.